jupyterhub_moss/spawner.py

```diff
@@ -297,4 +307,8 @@ def options_from_form(self, formdata: Dict[str, List[str]]) -> Dict[str, str]:
     async def submit_batch_script(self):
         self.log.info(f"Used environment: {self.user_options['environment_path']}")
         self.log.info(f"Used default URL: {self.default_url}")
+
+        # refresh environment to be kept in the job
+        self.req_keepvars = self._req_keepvars_default()
```

Comment:
What is the motivation to set it here?

It seems not related to the rest of the PR and sounds not always desired (e.g., if `req_keepvars` is set elsewhere).

Reply:
Thanks for the explanation! Looking at `req_keepvars` in `batchspawner`, it is not meant to be set from the config.

If possible, it would be better to avoid calling private methods, with e.g. [trait_defaults](https://github.com/ipython/traitlets/blob/65a4725a11c776054210583cf9da490e96bd2d23/traitlets/traitlets.py#L1740-L1746):
```suggestion
        # refresh environment to be kept in the job
        self.req_keepvars = self.trait_defaults("req_keepvars")
```

Reply:
good point, I was a bit reckless there :slightly_smiling_face: